File: 002_add_two_numbers/Solution.go
package main

import "fmt"

func main() {
	l1 := &ListNode{2, nil}
	l1.Next = &ListNode{4, nil}
	l1.Next.Next = &ListNode{3, nil}

	l2 := &ListNode{5, nil}
	l2.Next = &ListNode{6, nil}
	l2.Next.Next = &ListNode{4, nil}

	result := addTwoNumbers(l1, l2)
	for result != nil {
		fmt.Printf("%v ", result.Val)
		result = result.Next
	}
}

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	p, q := l1, l2
	dummyHead := &ListNode{0, nil}
	curr := dummyHead
	carry := 0

	for p != nil || q != nil {
		x, y := 0, 0
		if p != nil {
			x = p.Val
		} else {
			x = 0
		}

		if q != nil {
			y = q.Val
		} else {
			y = 0
		}

		sum := carry + x + y
		carry = sum / 10
		curr.Next = &ListNode{sum % 10, nil}
		curr = curr.Next

		if p != nil {
			p = p.Next
		}
		if q != nil {
			q = q.Next
		}
	}
	if carry > 0 {
		curr.Next = &ListNode{carry, nil}
	}

	return dummyHead.Next
}

File: 002_add_two_numbers/Solution.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        dummy_head = ListNode(0)
        p, q = l1, l2
        carry = 0
        curr = dummy_head

        while p or q:
            x = p.val if p else 0
            y = q.val if q else 0
            tmp_sum = carry + x + y
            carry = tmp_sum // 10
            curr.next = ListNode(tmp_sum % 10)
            curr = curr.next

            p = p.next if p else None
            q = q.next if q else None

        if carry > 0:
            curr.next = ListNode(carry)

        return dummy_head.next

